core/src/main/scala/kafka/server/ApiVersionManager.scala

```diff
@@ -73,18 +73,35 @@ class DefaultApiVersionManager(
 ) extends ApiVersionManager {
 
   override def apiVersionResponse(throttleTimeMs: Int): ApiVersionsResponse = {
+    val metadataVersion = metadataCache.metadataVersion()
     val supportedFeatures = features.supportedFeatures
     val finalizedFeatures = metadataCache.features()
     val controllerApiVersions = forwardingManager.flatMap(_.controllerApiVersions)
 
-    ApiVersionsResponse.createApiVersionsResponse(
+    val response = ApiVersionsResponse.createApiVersionsResponse(
         throttleTimeMs,
-        metadataCache.metadataVersion().highestSupportedRecordVersion,
+        metadataVersion.highestSupportedRecordVersion,
         supportedFeatures,
         finalizedFeatures.features.map(kv => (kv._1, kv._2.asInstanceOf[java.lang.Short])).asJava,
         finalizedFeatures.epoch,
         controllerApiVersions.orNull,
-        listenerType)
+        listenerType
+    )
+
+    // In ZK mode if the deployed software of the controller uses version 2.8 or above
+    // but the IBP is below 2.8, the controller does not assign topic ids. In this case,
+    // it should not advertise the AlterPartition API version 2 and above.
+    val alterPartitionApiVersion = response.apiVersion(ApiKeys.ALTER_PARTITION.id)
+    if (alterPartitionApiVersion != null) {
+      alterPartitionApiVersion.setMaxVersion(
+        if (metadataVersion.isTopicIdsSupported)
+          alterPartitionApiVersion.maxVersion()
+        else
+          1.toShort
+      )
+    }
```

Comment:
@hachikuji There is something that I missed in this PR. If the controller run 2.8 software or above but does not use IBP 2.8 or above yet, topic ids are not assigned. In other words, it is not safe to use the AlterPartition version 2 in this case even if the controller supports it. We don't really have a mechanism to do this at the moment so I have put the logic to do this here. We should definitely think about a better approach. What do you think?

Reply:
I have filed a JIRA for tracking this: https://issues.apache.org/jira/browse/KAFKA-13975. It seems to me that it would be better to do it separately.

Reply:
I don't follow. In this true for all RPC dealing with topic ids? The sender has IBP 2.8 but the receiver doesn't support IBP 2.8. I would think that in general the RPC receiver needs to allow for RPC from IBP versions that are greater than the local IBP.

Reply:
I recall there were some tricky cases when doing the upgrade to use TopicIds. It is possible for the controller to initialize on one of the nodes with the updated IBP and create TopicIds for all topics, but then change to a new node with a lower IBP. How does the controller handle the existence of TopicIds in the zk metadata if the IBP is below 2.8? At a quick glance, it looks like it would still parse it and load it into the `ControllerContext`. It seems ok in this scenario for the controller to accept `AlterPartition` with TopicIds. Are there any other scenarios we need to worry about?

Reply:
The other scenario that I was considering is the following:
* all brokers run software 2.8 or above
* controller upgrades to IBP 2.8 or above
* controller fails over to a node still on an IBP < 2.8 - topics with ID keep them here as you pointed out.
* new topics are created - those won't have topic ids

Is it safe? It seems to be OK because those new topics won't have a topic id so the AlterPartitionManager will downgrade to using version 1 in this case.

OK. I have convinced myself that we don't need this check after all. The AlterPartitionManager's logic to downgrade is sufficient to handle both cases. Thanks for the clarification.